ChangeLog.d/add-cipher-encrypt-only.txt

```diff
@@ -0,0 +1,7 @@
+Features
+   * Add support to remove xxx_setkey_dec and xxx_decrypt for cipher type of
```

Comment:
I think there are two big difference between `MBEDTLS_CIPHER_ENCRYPT_ONLY` (as it is now) and `MBEDTLS_CIPHER_LIGHT` (my current understanding of what it will be): first, the former affects other modules (`aes.c`, `aria.c`, `camellia.c` - removing `decrypt()` and `setkey_dec()` functions which those modules would normally have), while `MBEDTLS_CIPHER_LIGHT` is strictly about `cipher.c`.

I'll note there's a partially pre-existing naming issue here: in my ideal world, when an option's name starts with `MBEDTLS_CIPHER` it should only affect Cipher and modules that use it (PEM, etc.) but should not affect modules below it (AES, etc.). But that's not the case with all the `MBEDTLS_CIPHER_MODE_xxx` function which do affect `aes.c` etc. So, naming the option `MBEDTLS_CIPHER_ENCRYPT_ONLY` is consistent with `MBEDTLS_CIPHER_MODE_xxx` - it's an option that affects both `cipher.c` and multiple low-level modules (`aes.c`, `aria.c`, `camellia.c`). But that create a false parallel with `MBEDTLS_CIPHER_LIGHT`: they both start with `MBEDTLS_CIPHER` but that means different things.

The other big difference IMO is that `MBEDTLS_CIPHER_ENCRYPT_ONLY`, [as it is now](https://github.com/Mbed-TLS/mbedtls/pull/8124/files#diff-587ce90d84add2baf3903bb721c66b58d046f9a254738ed6095a4b53d726ffebR170) is essentially a negative option: when it's on, you get _less_ than when it's off. By contrast, `MBEDTLS_CIPHER_LIGHT` is a positive option: enabling it either has no effect (when `MBEDTLS_CIPHER_C` is already enabled) or gives you strictly more (a subset of `cipher.h` when `MBEDTLS_CIPHER_C` is disabled).

Btw, considering `ENCRYPT_ONLY` is a negative option, I'm not convinced we can implicitly enable it without breaking backwards compatibility.

Reply:
> Btw, considering ENCRYPT_ONLY is a negative option, I'm not convinced we can implicitly enable it without breaking backwards compatibility.

Yes, that's another important point. To clarify, the problem is that `MBEDTLS_CIPHER_C` has two meanings:

* The one we usually think about, which is: the application has requested `mbedtls_cipher_xxx` functions.
* One we tend to forget, which is: a promise that `mbedtls_cipher_xxx` functions work.

Application code can write
```
#if defined(MBEDTLS_CIPHER_C)
mbedtls_cipher_crypt(MBEDTLS_DECRYPT, ...);
#endif
```

If we add a negative option, then the application code would need to change to
```
#if defined(MBEDTLS_CIPHER_C) && !defined(MBEDTLS_CIPHER_NO_DECRYPT)
mbedtls_cipher_crypt(MBEDTLS_DECRYPT, ...);
#endif
```

“The application code would need to change” is a backward compatibility break.

As a guideline, if we're changing dependencies in TLS library or test code to be more general, that's ok. (For example: `MBEDTLS_AES_C` → `MBEDTLS_CIPHER_CAN_DO_AES` is ok, since `MBEDTLS_AES_C` implies `MBEDTLS_CIPHER_CAN_DO_AES`.) But if we're changing them to be more restrictive, that's potentially a problem.

`MBEDTLS_AES_C` is a promise that both AES encryption and AES decryption work, and with `MBEDTLS_CIPHER_C` it's a promise that `MBEDTLS_CIPHER_AES_128_ECB` (and 192, 256) works in both directions. I haven't looked much at the code in this pull request, but at this point I get the impression that it's breaking this promise.

Note that PSA has a different interface: `PSA_WANT_KEY_TYPE_AES` doesn't give anything useful on its own, only in combination. If you want AES-ECB, you need to also enable `PSA_WANT_ALG_ECB`. That's why it's easier to make encrypt-only builds as an optimization in PSA, whereas encrypt-only builds need an explicit configuration in the legacy interface.

Reply:
That's not how _some_ applications go about it. Others _do_ go about it like this.

A prime example is our TLS stack. If crypto changes require changes to the TLS stack to keep it working (as opposed to crypto changes _enabling_ changes to the TLS stack to make it work better), it's a sign that it would hurt other protocol stacks.

> From an application writer's perspective, MBEDTLS_AES_C is what I have to enable to get any AES functionality. If I don't need decrypt, because I'm only using forward-direction AES i.e. CTR, GCM, then there's currently no way to do it.

Right, and the proper way to solve that is a new option `MBEDTLS_AES_ENCRYPT`, which activates AES encryption but not decryption.

Reply:
> A prime example is our TLS stack

Sorry, I disagree here. Our TLS is a very bad example, because it is unique (well, along with `programs/` etc...) because it is in the same repository and configured in the very same files as the crypto primitives

Reply:
> because it is in the same repository and configured in the very same files as the crypto primitives

That makes it potentially a bad example of how you _should_ design a protocol stack and a crypto API. But it definitely is an example of how you _can plausibly_ design a protocol stack, given the existing mbedtls crypto API. That makes it a good guideline.

Reply:
> In my experience that's not how applications that use crypto libraries go about it. Rather, they have their code, and they configure the library so that they get the functionality they want.

That's probably true for embedded applications, but even though they're our primary target, they're not our only users. Non-embedded applications, most notably those that consume Mbed TLS as a dynamic library provided by the OS, tend to have a lot of checks about Mbed TLS version and build options and try to adapt. I think I saw a lot of that last time I looked at the source code of [hiawatha](https://www.hiawatha-webserver.org/about) for example.

> From an application writer's perspective, `MBEDTLS_AES_C` is what I have to enable to get any AES functionality.

Currently the meaning of `AES_C` is: enable this and you'll get the full primitive (encrypt and decrypt) plus whatever modes are enabled. It needs to keep doing that it all builds when users upgrade and keep their config files unchanged. That's our backwards compatibility promise.

Our only options for not breaking that promise are:
1. use a negative option but make it explicitly opt-in;
2. use a positive option.

I prefer a positive option, as it blends much better with existing options, and the PSA crypto config system where the users states their requirements and the system enables what's needed to meet them.

Reply:
(Note that as usual things would be much simpler if we already lived in a PSA-only world (including for crypto config). Applications could just either enable, or check for, `KEY_TYPE_AES` + `ALG_whatever`, those options would keep the same semantics, and we'd just include less code under the hood when possible. In that world, `AES_C` and `CIPHER_C` wouldn't public, and we wouldn't need to preserve their semantics. But that's just not the world we live in right now.)